pulsar/internal/retry.go

```diff
@@ -0,0 +1,62 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+type OpFn[T any] func() (T, error)
+
+// Retry the given operation until the returned error is nil or the context is done.
+func Retry[T any](ctx context.Context, op OpFn[T], nextDuration func(error) time.Duration) (T, error) {
+	var (
+		timer *time.Timer
+		res   T
+		err   error
+	)
+
+	cleanTimer := func() {
+		if timer != nil && !timer.Stop() {
+			<-timer.C
+		}
+	}
+	defer cleanTimer()
+
+	for {
+		res, err = op()
+		if err == nil {
+			return res, nil
+		}
+
+		duration := nextDuration(err)
+		if timer == nil {
+			timer = time.NewTimer(duration)
+		} else {
+			timer.Reset(duration)
```

Comment:
According to the [go doc of the timer](https://pkg.go.dev/time#Timer.Reset):

> If a program has not yet received a value from t.C, however, the timer must be stopped and—if Stop reports that the timer expired before being stopped—the channel explicitly drained

We should stop the timer and drain the channel before resetting the timer.

Reply:
> // If a program has already received a value from t.C, the timer is known
> // to have expired and the channel drained, so t.Reset can be used directly.

The `timer.C` has been consumed, so we can use `timer.Reset` directly.